patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Terraria.WorldBuilding;
+
+namespace Terraria.ModLoader
+{
+	public static class BiomeLoader
+	{
+		internal static readonly IList<ModBiome> biomes = new List<ModBiome>();
+
+		private class HookList
+		{
+			public int[] arr;
+			public readonly MethodInfo method;
+
+			public HookList(MethodInfo method) {
+				this.method = method;
+			}
+		}
+
+		private static List<HookList> hooks = new List<HookList>();
+
+		private static HookList AddHook<F>(Expression<Func<ModBiome, F>> func) {
+			var hook = new HookList(ModLoader.Method(func));
+			hooks.Add(hook);
+			return hook;
+		}
+
+		internal static void Add(ModBiome biome) {
+			biome.index = biomes.Count;
+			biomes.Add(biome);
+		}
+
+		internal static void RebuildHooks() {
+			foreach (var hook in hooks) {
+				hook.arr = ModLoader.BuildGlobalHook(biomes, hook.method).Select(p => p.index).ToArray();
+			}
+		}
+
+		// Internal boilerplate
+
+		internal static void SetupPlayer(Player player) {
+			player.modBiomeFlags = new bool[biomes.Count];
+		}
+
+		public static void UpdateBiomes(Player player) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				biomes[i].Update(player, ref player.modBiomeFlags[i]);
+			}
+		}
+
+		public static bool CustomBiomesMatch(Player player, Player other) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				if (player.modBiomeFlags[i] ^ other.modBiomeFlags[i])
+					return false;
+			}
+			return true;
+		}
+
+		public static void CopyCustomBiomesTo(Player player, Player other) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				other.modBiomeFlags[i] = player.modBiomeFlags[i];
+			}
+		}
+
+		public static void SendCustomBiomes(Player player, BinaryWriter writer) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				writer.Write(player.modBiomeFlags[i]);
```

Comment:
bit packing please. Honestly the whole `modBiomeFlags` should probably be https://docs.microsoft.com/en-us/dotnet/api/system.collections.bitarray?view=netframework-4.5 instead of `bool[]`

Note that `BitArray` is optimised for performance. One day we might consider (or get yorai to consider) them for ID.Sets as well

Reply:
Hmm, well for `BitArray`, the internal backing field isn't exposed at all (`m_array`). When I use `BitArray`, I just use reflection to retrieve it, but I don't think that's the right way to serialize it, is it?